saturn_client/core.py

```diff
@@ -0,0 +1,161 @@
+
+        :param name: Name of project.
+        :param description: Short description of the project (less than 250 characters).
+        :param image_uri: Location of the image. Example:
+            485185227295.dkr.ecr.us-east-1.amazonaws.com/saturn-dask:2020.12.01.21.10
+        :param start_script: Script that runs on start up. Examples: "pip install dask"
+        :param environment_variables: Env vars expressed as a dict. The names will be
+            coerced to uppercase.
+        :param working_dir: Location to use as working directory. Example: /home/jovyan/project
+        :param workspace_settings: Setting for the jupyter associated with the project.
+            These include: `size`, `disk_space`, `auto_shutoff`, and `start_ssh`. The options
+            for these are available from `conn.options`.
+        """
+        url = urljoin(self.url, "api/projects")
+
+        errors = []
+        if environment_variables:
+            environment_variables = "\n".join(
+                f"{k.upper()}={v}" for k, v in environment_variables.items()
+            )
+
+        workspace_kwargs = {}
+        if workspace_settings:
+            for k, v in workspace_settings.items():
+                workspace_keys = ["size", "auto_shutoff", "disk_space", "start_ssh"]
+                if k in workspace_keys:
+                    if k == "start_ssh":
+                        if not isinstance(v, bool):
+                            errors.append(f"{k} must be set to a boolean if defined.")
+                    else:
+                        options = (
+                            list(self.options["sizes"].keys()) if k == "size" else self.options[k]
+                        )
+                        if v not in options:
+                            errors.append(
+                                f"Proposed {k}: {v} is not a valid option. Options are: {options}."
+                            )
+                    workspace_kwargs[f"jupyter_{k}"] = v
+                else:
+                    errors.append(
+                        f"{k} is not a valid workspace_setting. "
+                        f"Supported workspace_settings are {workspace_keys}."
+                    )
+
+        if len(errors) > 0:
+            raise ValueError(errors)
+
+        project_config = {
+            "name": name,
+            "description": description,
+            "image": image_uri,
+            "start_script": start_script,
+            "environment_variables": environment_variables,
+            "working_dir": working_dir,
+            **workspace_kwargs,
+        }
+        # only send kwargs that are explicitly set by user
+        project_config = {k: v for k, v in project_config.items() if v is not None}
+
+        response = requests.post(
+            url,
+            data=json.dumps(project_config),
+            headers=self.settings.headers,
+        )
+        if not response.ok:
+            raise ValueError(response.json()["message"])
```

Comment:
I don't think a `ValueError` here is appropriate. I think it would be more accurate to do

```python
response.raise_for_status()
```

That will return a `requests.exceptions.HTTPError, and it'll include the response code and message in it`. That's really valuable, because it can be used by client code.

Reply:
Oh nice!! I didn't know about that one.

Reply:
hmm doesn't quite work since we return the error message in the json as "message"

Reply:
sad, ok. Ignore this one then.